### PR TITLE
Add ccstd::monostate for the first type of ccstd::variant.

### PR DESCRIPTION
**Pull request by dumganhar**
```c++
#include <iostream>
#include <variant>
#include <string>

using MacroValue = std::variant<int32_t, bool, std::string>;

int main(int argc, const char * argv[]) {
    // insert code here...
    std::cout << "Hello, World!\n";
    MacroValue v;
    if (std::holds_alternative<int32_t>(v)) {
        int32_t a = std::get<int32_t>(v);
        printf("it's a int32_t type, the value is %d\n", a); // Will Go here and print 0
    }

    return 0;
}
```

std::variant uses the first element as its default type, that will cause some wrong behaviors. So add `std::monostate` as its first element, and always check the invalid state (std::monostate) while invoking `std::visit`.

Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/base/TemplateUtils.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/****************************************************************************
+ Copyright (c) 2010 cocos2d-x.org
+ Copyright (c) 2013-2016 Chukong Technologies Inc.
+ Copyright (c) 2017-2022 Xiamen Yaji Software Co., Ltd.
```
Comment (minggo) [inline native/cocos/base/TemplateUtils.h]: error date

Comment (minggo) [reply native/cocos/base/TemplateUtils.h]: and only `Xiamen Yaji` needed

**Review by minggo — APPROVED** [review]

File: native/cocos/base/TemplateUtils.h (hunk 1 above)
Comment (dumganhar) [reply native/cocos/base/TemplateUtils.h]: OK

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Comment by LIUYIa** [comment]
@cocos-robot run test cases

**Pull request merged**